SCORER.md

```diff
@@ -21,11 +25,231 @@ You may need to build your own scorer - your own _language model_ if:
+
+1. Having, or preparing, a text file (in `.txt` or `.txt.gz` format), with one phrase or word on each line. If you are training a speech recognition model for a particular domain - such as technical words, medical transcription, agricultural terms etc, then they would appear in the text file.
+
+2. Using the `lm_optimizer.py` script with _checkpoints_ from a trained model to generate values for the parameters `--default_alpha` and `--default_beta` that are used by the `generate_scorer_package` script.
+
+3. Using the `generate_lm.py` script which is distributed with DeepSpeech, along with the text file, to create two files, called `lm.binary` and `vocab-500000.txt`.
+
+4. Downloading the prebuilt `native_client` from the DeepSpeech repository on GitHub, and using the `generate_scorer_package` to create a `kenlm.scorer` file.
+
+5. Using the `kenlm.scorer` file as the _external_scorer_ passed to `DeepSpeech.py` during training.
+
+In the following example we will create a custom external scorer file for Bahasa Indonesia (BCP47: `id-ID`).
+
+#### Preparing the text file
+
+This is straightforward. In this example, we will use a file called `indonesian-sentences.txt`. This file has been created by copying phrases from the `train.tsv`, `validate.tsv` and `test.tsv` files in the Common Voice Indonesian dataset.
+
+```
+~/deepspeech-data$ ls cv-corpus-6.1-2020-12-11/id
+total 6288
+   4 drwxr-xr-x 3 root root    4096 Feb 24 19:01 ./
+   4 drwxr-xr-x 4 root root    4096 Feb 11 07:09 ../
+1600 drwxr-xr-x 2 root root 1638400 Feb  9 10:43 clips/
+ 396 -rwxr-xr-x 1 root root  401601 Feb  9 10:43 dev.tsv
+ 104 -rwxr-xr-x 1 root root  103332 Feb  9 10:43 invalidated.tsv
+1448 -rwxr-xr-x 1 root root 1481571 Feb  9 10:43 other.tsv
+  28 -rwxr-xr-x 1 root root   26394 Feb  9 10:43 reported.tsv
+ 392 -rwxr-xr-x 1 root root  399790 Feb  9 10:43 test.tsv
+ 456 -rwxr-xr-x 1 root root  465258 Feb  9 10:43 train.tsv
+1848 -rwxr-xr-x 1 root root 1889606 Feb  9 10:43 validated.tsv
+```
+
+The `indonesian-sentences.txt` file is stored on the local filesystem in the `deepspeech-data` directory so that the Docker container can access it.
+
+```
+~/deepspeech-data$ ls | grep indonesian-sentences
+ 476 -rw-rw-r--  1 root root  483481 Feb 24 19:02 indonesian-sentences.txt
+```
+
+#### Using `lm_optimizer.py` with the text file to generate values for the parameters `--default_alpha` and `--default_beta` that are used by the `lm_optimizer.py` script
+
+The `lm_optimizer.py` script is located in the `DeepSpeech` directory if you have set up your [environment][ENVIRONMENT.md] as outlined in the PlayBook.
+
+```
+root@57e6bf4eeb1c:/DeepSpeech# ls | grep lm_optimizer.py
+lm_optimizer.py
+```
+
+This script takes a text file, e.g. `indonesian-sentences.txt`, and a set of _checkpoints_, and determines the optimal `--default_alpha` and `--default_beta` values. These are then passed to `generate_lm.py` in the next step.
```

Comment:
No, the `lm_optimizer` works out of CSV data:
```
		python -u lm_optimizer.py \
			--show_progressbar True \
			--train_cudnn True \
			--alphabet_config_path /mnt/models/alphabet.txt \
			--scorer_path /mnt/lm/kenlm.scorer \
			--feature_cache /mnt/sources/feature_cache \
			--test_files ${all_test_csv} \
			--test_batch_size ${BATCH_SIZE} \
			--n_hidden ${N_HIDDEN} \
			--lm_alpha_max ${LM_ALPHA_MAX} \
			--lm_beta_max ${LM_BETA_MAX} \
			--n_trials ${LM_N_TRIALS} \
			--checkpoint_dir /mnt/checkpoints/
```

And specifically expects `lm_alpha_max`, `lm_beta_max` as well as `n_trials` (default values might be too wide).

Reply:
Great pickup, thank you.